docs/proposals/fluentd-limit-requests-to-apiserver.md

```diff
@@ -0,0 +1,63 @@
+# Modifications to Fluentd Metadata Retrieval Proposal
+
+##Motivation
+In the current design (v3.4) of the EFK stack, there is a Fluentd pod deployed to each Openshift node. Fluentd is responsible for reading container logs, adding pod metadata (e.g. labels, namespace) to individual log entries, and sending the logs to the Elasticsearch cluster. The pod metadata is sourced by watching the Openshift API server for Pod information via the [Kubernetes Metadata Filter](https://github.com/fabric8io/fluent-plugin-kubernetes_metadata_filter)[2]. Testing at larger scales of Openshift clusters has shown issues where Fluentd is unable to connect to the API server because it is limited by the **maxRequestsInFlight** configuration setting [1].  This results in logs not being shipped to the aggregator and thus being unavailable.
+
+##Proposal - Short Term (3.4)
+1. Leave the value as-is and note that it must be adjusted to account for # of nodes plus some headroom (swag @ 10%?).
+1. Increase the default to our current maximum supported node count, plus some headroom.
+1. Increasing maxRequestsInflight upstream
+
+##Proposal - Long Term
+###Option 1 - Admission Controller
+1. Create a simple admission controller which simply embeds the necessary meta-data as env-variables into the pods on submission.  In this way every pod that is run on the system will have its data locally.
+1. Mount the docker socket into the Fluentd pod such that it can query containers env-variables.
```

Comment:
It makes the most sense to decentralize the metadata, meaning ensure that the necessary data is available on the node and no new per-node master API request is necessary to augment metadata. That will scale indefinitely and minimize complexity. Also it will support sending enriched logs directly from fluentd to other sources besides our blessed ES - an external ES or fluentd as some customers are requesting.

We should have the necessary metadata already in the kubelet. All we need to do is make it available to the node's fluentd pod (authentication needed and possibly changes to kubelet API). I would expect the kubelet to even get updates if the metadata changes.

The only real downside of decentralizing is that we're no longer talking to the authoritative source of information, so we risk having stale data. In practice, this isn't going to happen much - there's not much of a use case for changing a pod once it's been scheduled to a node, and if logs were given (for a brief period of a minute or two) outdated labels, I can live with that.